### PR TITLE
[GEN-660] Session Revamp (New Collection)

### PR DESCRIPTION
**Pull request by abhipi**
- tapActions and Record/Backend/logic were changed to create a new Session or update the Current session.
- ListHistory was changed to remve reversing the list (is now fetched from firebase in order).
- New SessionsApi created and TapsApi updated to remove unused functions.
- SessionsTapHistory updated to allow tagging the current session.
- SessionDetail, Edit_Session updated to deal with the new data (they load data independently now so can implement future reloading).
- Session Item loads faster as only required data is passed in (Session Collection Object data ONLY!).

https://github.com/nagimonyc/trk-app/assets/75244191/8f4c6e98-ef66-48a0-9e29-fc1767db7158

### CODE REVIEW

File: trk/src/Actions/tapActions.js [hunk 1]
```diff
@@ -69,7 +103,22 @@ export const processClimbId = (climbId, currentUser, role) => {
                         // Getting the Error details.
                         console.error('Error calling function:', error);
                     });
+            } else {
+              //If the tap is not the start of a session, place it in climbs of that session (fetch by expiryTime)
+              const lastSessionSnapshot = await SessionsApi().getLastUserSession(currentUser.uid);
+              if (lastSessionSnapshot.empty) {
+                  //Error, if it is not the start of a session, there should be data here
+                  console.error("Session doesn't exist, but current tap is not the start of the session.");
```
Comment (redpepper-nag) [inline trk/src/Actions/tapActions.js:111]: Smart error catching

**Review by redpepper-nag — APPROVED** [review]

File: trk/src/api/TapsApi.js [hunk 2]
```diff
@@ -44,50 +36,34 @@ function TapsApi() {
...
+                }
+                if (featuredClimbCopy === tapId) {
+                    featuredClimbCopy = climbsArray.length > 0 ? climbsArray[0] : ''; // Set to first element or null if climbsArray is empty
+                }
+                if (climbsArray.length == 0) {
+                    //session is empty, archive it
+                    await SessionsApi().updateSession(sessionSnapshot.docs[0].id, {archived: true});
+                } else {
+                    //update session with the new climbsarray and featuredClimb
+                    await SessionsApi().updateSession(sessionSnapshot.docs[0].id, {climbs: climbsArray, featuredClimb: featuredClimbCopy});
```
Comment (VGmt) [inline trk/src/api/TapsApi.js:66]: amazing

Comment (VGmt) [reply trk/src/api/TapsApi.js:66]: good job covering edge cases (array length check & assignment to archived if empty)

File: trk/src/api/TapsApi.js [hunk 3]
```diff
@@ -125,39 +101,8 @@ function TapsApi() {
...
-    function getExpiredTaps(userId) {
-        //console.log('User ID: ', userId);
-        //console.log('Fetching Expired Climbs....');
-        return ref
-            .where('user', '==', userId)
-            .where("expiryTime", '<=', firebase.firestore.Timestamp.now())
-            .orderBy("expiryTime", "desc")
-            .orderBy('timestamp', 'desc') //For corrrect ordering
+            .orderBy("expiryTime")
+            .orderBy('timestamp', 'desc') //To avoid rerendering on reversing the climb list
```
Comment (VGmt) [inline trk/src/api/TapsApi.js:105]: smart

File: trk/src/Components/SessionItem.js [hunk 4]
```diff
@@ -46,47 +48,58 @@ const styles = StyleSheet.create({
     },
 });
 
-const SessionItem = ({ data, title, renderItem, keyExtractor, isHighlighted }) => {
+//Only Session Item relevant data is calculated here now, rest of the data is passed into Detail and Edit.
```
Comment (VGmt) [inline trk/src/Components/SessionItem.js:51]: ofc.. amazing!

**Review by VGmt — APPROVED** [review]

**Pull request merged**